Выполнение домашнего задание по 2му спринту

Nihrom77 opened the pull request:
Тесты для блока видео "Тестирование".

Commit:
Добавил тесты для пакета QA.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/93/

Commit:
Реализовал задания по теме ООП.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/104/

Nihrom77 commented:
Реализовал задания по теме ООП.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/105/

Commit:
Реализовал задания по теме IO.
Написал тест, проверяющий операции по архивации / разархивированию.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/112/

Commit:
Реализовал задания по теме NIO.
Добавил тест, проверяющий корректность поиска подстроки.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/113/

Commit:
Исправил замечания от преподавателя.
Переписал класс HrDepartment, выделил функциональность записи/чтения писем в отдельный сервис. Чтобы не нужно было через рефлексию читать приватные поля.
Разделил тест, проверяющий отправку и чтение входящего запроса в HrDepartment.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/114/

Nihrom77 commented:
Комментарий преподавателя на платформе:

> Юнит-тест должен проверять только один метод.
> Не должно быть вызовов receiveRequest и processNextRequest внутри одного теста.
> 
> + стоит добавить дополнительные проверки корректной работы методов (присутствуют только проверки на исключения, но ведь метод может работать неправильно и не выбрасывая их)

Думаю, что все же юнит-тест должен проверять "минимальную функциональность", не всегда это один метод. В классе  HrDepartment методы receiveRequest и processNextRequest лишь меняют внутренне состояние класса, которое скрыто за приватными полями. Я не стал через рефлексию проверять состояние приватный переменных в тестах, потому что придерживаюсь правила - если в тестах требуется проверять приватные методы/поля, значит класс спроектирован плохо. Вынес внутренне состояние в отдельный класс HrDepartmentExchange с публичными методами, вызовы которых и проверяю в тестах.

Commit:
Переписал тест, проверяющий класс HrDepartment с использованием рефлексии для проверки внутреннего состояния класса после вызовов его методов.

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/132/

atitorenko commented on oop/src/main/kotlin/ru/sber/oop/DataClasses.kt:
```diff
@@ -2,10 +2,35 @@ package ru.sber.oop
+    /**
+     * Переопределяем equals, чтобы учитывать все поля
+     */
+    override fun equals(other: Any?): Boolean {
```
неправильно определена функция. она не учитывает поля класса и возвращает false даже если объекты имеют одинаковые поля

atitorenko commented on oop/src/main/kotlin/ru/sber/oop/Inheritance.kt:
```diff
@@ -1,13 +1,46 @@
+    val monster: Monster = Goblin("особо уродливый")
+}
+
+class TownSquare : Room(name = "Town Square", size = 1000) {
```
класс TownSquare по умолчанию final. его следует сделать open, чтобы имел смысл запрет переопределения функции load

Commit:
Для модуля ООП переписал классы. Убрал main() метод, добавил тесты по каждому пункту заданий.
Поправил ошибки в методе equals()

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/163/

atitorenko commented on oop/src/main/kotlin/ru/sber/oop/DataClasses.kt:
```diff
@@ -2,10 +2,18 @@ package ru.sber.oop
+        return super.equals(other)
+    }
+
+    override fun hashCode(): Int {
```
hashcode должен быть одинаковым если объекты equals true

Nihrom77 replied:
Поправил

Commit:
Исправил функцию hashCode, чтобы сохранить контракт - если объекты равны через equals, то у них равны hashCode()

shariktlt-devops reviewed (approved):
:ok: build ok, see: https://jenkins.shariktlt.ru/job/KotlinSchool/job/PR_Builder/192/
